File: shared/state.ts
import { Shape } from 'shape';
import * as uuid from 'uuid';
import { VERSION } from './constants';
import { SupportedLanguages } from './languages';
import { mergeDeep, setPath } from './object';

export class PanelResult {
  exception?: any;
  value?: Array<any>;
  preview: string;
  stdout: string;
  shape: Shape;
  arrayCount: null | number;
  size: number;
  contentType: string;
  elapsed?: number;
  lastRun?: Date;
  loading: boolean;

  constructor(defaults: Partial<PanelResult> = {}) {
    this.lastRun = defaults.lastRun || null;
    this.loading = defaults.loading || false;
    this.stdout = defaults.stdout || '';
    this.shape = defaults.shape || { kind: 'unknown' };
    this.preview = defaults.preview || '';
    this.size = defaults.size || 0;
    this.contentType = defaults.contentType || 'unknown';
    this.elapsed = defaults.elapsed || 0;
    this.value = defaults.value || undefined;
    this.exception = defaults.exception || undefined;
  }

  static fromJSON(raw: any): PanelResult {
    raw = raw || {};
    raw.loading = false;
    const prm = mergeDeep(new PanelResult(), raw);
    prm.lastRun =
      typeof raw.lastRun === 'string'
        ? new Date(raw.lastRun)
        : raw.lastRun || prm.lastRun;
    return prm;
  }
}

export class Encrypt {
  value: string;
  encrypted: boolean;

  constructor(value: string = '') {
    this.value = value;
    this.encrypted = false;
  }

  static Encrypted(value: string = '') {
    const e = new Encrypt(value);
    e.encrypted = true;
    return e;
  }
}

export type ServerInfoType = 'ssh-agent' | 'password' | 'private-key';

export class ServerInfo {
  name: string;
  address: string;
  port: number;
  type: ServerInfoType;
  username: string;
  password_encrypt: Encrypt;
  privateKeyFile: string;
  passphrase_encrypt: Encrypt;
  id: string;

  constructor(panel: Partial<ServerInfo> = {}) {
    this.type = panel.type || 'private-key';
    this.name = panel.name || 'Untitled Server';
    this.address = panel.address || '';
    this.port = panel.port || 22;
    this.username = panel.username || '';
    this.password_encrypt = panel.password_encrypt || new Encrypt('');
    this.privateKeyFile = panel.privateKeyFile || '';
    this.passphrase_encrypt = panel.passphrase_encrypt || new Encrypt('');
    this.id = uuid.v4();
  }

  static fromJSON(raw: any): ServerInfo {
    raw = raw || {};
    return mergeDeep(new ServerInfo(), raw);
  }
}

export type ConnectorInfoType = 'database' | 'http';

export class ConnectorInfo {
  name: string;
  type: ConnectorInfoType;
  id: string;
  serverId?: string;

  constructor(type?: ConnectorInfoType, name?: string, serverId?: string) {
    this.name = name || 'Untitled Data Source';
    this.type = type || 'database';
    this.serverId = serverId;
    this.id = uuid.v4();
  }

  static fromJSON(raw: any): ConnectorInfo {
    raw = raw || {};
    // Migrate panel name from sql to database
    if (raw.type === 'sql') {
      raw.type = 'database';
      raw.database = raw.sql;
      delete raw.sql;
    }
    const ci = mergeDeep(new ConnectorInfo(), raw);

    switch (raw.type) {
      case 'database':
        return mergeDeep(new DatabaseConnectorInfo(), ci);
      case 'http':
        return mergeDeep(new HTTPConnectorInfo(), ci);
    }
    return ci;
  }
}

export type HTTPConnectorInfoMethod =
  | 'GET'
  | 'HEAD'
  | 'PUT'
  | 'POST'
  | 'DELETE';

export class ContentTypeInfo {
  type: string;
  customLineRegexp: string;

  constructor(type?: string, customLineRegexp?: string) {
    this.type = type || '';
    this.customLineRegexp = customLineRegexp || '';
  }

  static fromJSON(raw: any): ContentTypeInfo {
    return mergeDeep(new ContentTypeInfo(), raw);
  }
}

export class HTTPConnectorInfo extends ConnectorInfo {
  http: {
    headers: Array<{ value: string; name: string }>;
    url: string;
    method: HTTPConnectorInfoMethod;
    contentTypeInfo: ContentTypeInfo;
  };

  constructor(
    name?: string,
    url?: string,
    headers: Array<{ value: string; name: string }> = [],
    method?: HTTPConnectorInfoMethod,
    contentTypeInfo?: ContentTypeInfo
  ) {
    super('http', name);
    this.http = {
      headers,
      url: url || '',
      method: method || 'GET',
      contentTypeInfo: contentTypeInfo || new ContentTypeInfo(),
    };
  }
}

export type SQLConnectorType =
  | 'postgres'
  | 'mysql'
  | 'sqlite'
  | 'oracle'
  | 'sqlserver'
  | 'presto'
  | 'clickhouse'
  | 'snowflake'
  | 'cockroach'
  | 'timescale'
  | 'crate'
  | 'yugabyte'
  | 'quest'
  | 'bigquery'
  | 'cassandra'
  | 'scylla'
  | 'athena';

export type DatabaseConnectorInfoType =
  | SQLConnectorType
  | 'google-sheets'
  | 'airtable'
  | 'mongo'
  | 'elasticsearch'
  | 'splunk'
  | 'prometheus'
  | 'influx'
  | 'influx-flux';

export class DatabaseConnectorInfo extends ConnectorInfo {
  database: {
    type: DatabaseConnectorInfoType;
    database: string;
    username: string;
    password_encrypt: Encrypt;
    address: string;
    apiKey_encrypt: Encrypt;
    extra: Record<string, string>;
  };

  constructor(
    panel: Partial<DatabaseConnectorInfo['database'] & { name: string }> = {}
  ) {
    super('database', panel.name || '');
    this.database = {
      type: panel.type || 'postgres',
      database: panel.database || '',
      username: panel.username || '',
      password_encrypt: panel.password_encrypt || new Encrypt(''),
      address: panel.address || '',
      extra: panel.extra || {},
      apiKey_encrypt: panel.apiKey_encrypt || new Encrypt(''),
    };
  }
}

export type PanelInfoType =
  | 'table'
  | 'http'
  | 'graph'
  | 'program'
  | 'literal'
  | 'database'
  | 'file'
  | 'filagg';

export class PanelInfo {
  content: string;
  type: PanelInfoType;
  name: string;
  id: string;
  serverId: string;
  resultMeta: PanelResult;
  lastEdited: Date;
  pageId: string;

  constructor(
    type: PanelInfoType,
    pageId: string,
    name?: string,
    content?: string
  ) {
    this.content = content || '';
    this.type = type;
    this.name = name || '';
    this.id = uuid.v4();
    this.resultMeta = new PanelResult();
    this.lastEdited = new Date();
    this.pageId = pageId;
  }

  static fromJSON(raw: any): PanelInfo {
    raw = raw || {};

    // Migrate panel name from sql to database
    if (raw.type === 'sql') {
      raw.type = 'database';
      raw.database = raw.sql;
      delete raw.sql;
    }

    let pit: PanelInfo = mergeDeep(
      new PanelInfo(raw.type || 'literal', raw.pageId),
      raw
    );

    pit.lastEdited =
      (typeof raw.lastEdited === 'string'
        ? new Date(raw.lastEdited)
        : raw.lastEdited) || pit.lastEdited;

    switch (pit.type) {
      case 'table':
        pit = mergeDeep(new TablePanelInfo(pit.pageId), pit);
        break;
      case 'http':
        pit = mergeDeep(new HTTPPanelInfo(pit.pageId), pit);
        break;
      case 'graph':
        pit = mergeDeep(new GraphPanelInfo(pit.pageId), pit);
        break;
      case 'program':
        pit = mergeDeep(new ProgramPanelInfo(pit.pageId), pit);
        break;
      case 'literal':
        pit = mergeDeep(new LiteralPanelInfo(pit.pageId), pit);
        break;
      case 'database':
        pit = mergeDeep(new DatabasePanelInfo(pit.pageId), pit);
        break;
      case 'file':
        pit = mergeDeep(new FilePanelInfo(pit.pageId), pit);
        break;
      case 'filagg':
        pit = mergeDeep(new FilterAggregatePanelInfo(pit.pageId), pit);
        break;
    }

    pit.resultMeta = PanelResult.fromJSON(raw.resultMeta);
    return pit;
  }
}

export class ProgramPanelInfo extends PanelInfo {
  program: {
    type: SupportedLanguages;
  };

  constructor(
    pageId: string,
    {
      name,
      type,
      content,
    }: Partial<
      ProgramPanelInfo['program'] & { content: string; name: string }
    > = {}
  ) {
    super('program', pageId, name || '', content || '');
    this.program = {
      type: type || 'python',
    };
  }
}

export interface GraphField {
  field: string;
  label: string;
}

export type GraphPanelInfoType = 'bar' | 'pie' | 'line';

export type PanelInfoWidth = 'small' | 'medium' | 'large';

export class GraphPanelInfo extends PanelInfo {
  graph: {
    panelSource: string;
    ys: Array<GraphField>;
    x: string;
    uniqueBy: string;
    type: GraphPanelInfoType;
    width: PanelInfoWidth;
    colors: {
      unique: boolean;
    };
  };

  constructor(
    pageId: string,
    defaults: Partial<
      GraphPanelInfo['graph'] & { content: string; name: string }
    > = {}
  ) {
    super('graph', pageId, defaults.name, defaults.content);
    this.graph = {
      panelSource: defaults.panelSource || '',
      x: defaults.x || '',
      uniqueBy: defaults.uniqueBy || '',
      ys: defaults.ys || [],
      type: defaults.type || 'line',
      width: defaults.width || 'small',
      colors: defaults.colors || {
        unique: false,
      },
    };
  }
}

export type TimeSeriesRelativeTimes =
  | 'last-5-minutes'
  | 'last-15-minutes'
  | 'last-30-minutes'
  | 'last-hour'
  | 'last-3-hours'
  | 'last-6-hours'
  | 'last-12-hours'
  | 'last-day'
  | 'last-3-days'
  | 'last-week'
  | 'last-2-weeks'
  | 'last-month'
  | 'last-2-months'
  | 'last-3-months'
  | 'last-6-months'
  | 'last-year'
  | 'last-2-years'
  | 'all-time';

export type TimeSeriesFixedTimes =
  | 'this-hour'
  | 'previous-hour'
  | 'today'
  | 'yesterday'
  | 'week-to-date'
  | 'previous-week'
  | 'month-to-date'
  | 'previous-month'
  | 'quarter-to-date'
  | 'previous-quarter'
  | 'year-to-date'
  | 'previous-year';

export type TimeSeriesRange = {
  field: string;
  sortOn?: string;
  sortAsc?: boolean;
} & (
  | {
      rangeType: 'absolute';
      begin_date: Date;
      end_date: Date;
    }
  | {
      rangeType: 'relative';
      relative: TimeSeriesRelativeTimes;
    }
  | {
      rangeType: 'fixed';
      fixed: TimeSeriesFixedTimes;
    }
);

export class DatabasePanelInfo extends PanelInfo {
  database: {
    connectorId?: string;
    range: TimeSeriesRange;
    table: string;
    step: number;
    extra: Record<string, string>;
  };

  constructor(
    pageId: string,
    panel: Partial<
      DatabasePanelInfo['database'] & { content: string; name: string }
    > = {}
  ) {
    super('database', pageId, panel.name || '', panel.content || '');
    this.database = {
      connectorId: panel.connectorId || '',
      range: panel.range || {
        field: '',
        rangeType: 'relative',
        relative: 'last-hour',
      },
      table: panel.table || '',
      step: panel.step || 60,
      extra: panel.extra || {},
    };
  }
}

export class HTTPPanelInfo extends PanelInfo {
  http: HTTPConnectorInfo;

  constructor(
    pageId: string,
    name?: string,
    http?: HTTPConnectorInfo,
    content?: string
  ) {
    super('http', pageId, name, content);
    this.http = http || new HTTPConnectorInfo();
  }
}

export interface TableColumn {
  label: string;
  field: string;
}

export class TablePanelInfo extends PanelInfo {
  table: {
    columns: Array<TableColumn>;
    panelSource: string;
    width: PanelInfoWidth;
    rowNumbers: boolean;
  };

  constructor(
    pageId: string,
    defaults: Partial<
      TablePanelInfo['table'] & { content: string; name: string }
    > = {}
  ) {
    super('table', pageId, defaults.name, defaults.content);
    this.table = {
      columns: defaults.columns || [],
      panelSource: defaults.panelSource || '',
      width: defaults.width || 'small',
      rowNumbers: defaults.rowNumbers || true,
    };
  }
}

export type AggregateType =
  | 'none'
  | 'count'
  | 'sum'
  | 'average'
  | 'min'
  | 'max';

export class FilterAggregatePanelInfo extends PanelInfo {
  filagg: {
    panelSource: string;
    filter: string;
    range: TimeSeriesRange;
    aggregateType: AggregateType;
    groupBy: string;
    aggregateOn: string;
    sortOn: string;
    sortAsc: boolean;
    windowInterval: string;
    limit: number;
  };

  constructor(
    pageId: string,
    panel: Partial<FilterAggregatePanelInfo['filagg'] & { name: string }> = {}
  ) {
    super('filagg', pageId, panel.name || '', '');
    this.filagg = {
      panelSource: panel.panelSource || '',
      filter: panel.filter || '',
      aggregateType: panel.aggregateType || 'none',
      groupBy: panel.groupBy || '',
      aggregateOn: panel.aggregateOn || '',
      sortOn: panel.sortOn || '',
      sortAsc: panel.sortAsc || false,
      limit: panel.limit || 100,
      windowInterval: panel.windowInterval || '',
      range: panel.range || {
        field: '',
        rangeType: 'relative',
        relative: 'last-hour',
      },
    };
  }
}

export class FilePanelInfo extends PanelInfo {
  file: {
    contentTypeInfo: ContentTypeInfo;
    name: string;
    content: ArrayBuffer;
  };

  constructor(
    pageId: string,
    panel: Partial<FilePanelInfo['file'] & { panelName: string }> = {}
  ) {
    super('file', pageId, panel.panelName, '');
    this.file = {
      name: panel.name || '',
      content: panel.content || new ArrayBuffer(0),
      contentTypeInfo: panel.contentTypeInfo || new ContentTypeInfo(),
    };
  }
}

export class LiteralPanelInfo extends PanelInfo {
  literal: {
    contentTypeInfo: ContentTypeInfo;
  };

  constructor(
    pageId: string,
    {
      name,
      content,
      contentTypeInfo,
    }: Partial<
      LiteralPanelInfo['literal'] & { content: string; name: string }
    > = {}
  ) {
    super('literal', pageId, name || '', content || '');
    this.literal = {
      contentTypeInfo: contentTypeInfo || new ContentTypeInfo(),
    };
  }
}

export class DashboardPanel {
  id: string;
  panelId: string;
  x: number;
  y: number;
  width: number;
  height: number;
}

export type ProjectPageVisibility = 'no-link' | 'private-link' | 'public-link';

export class Dashboard {
  id: string;
  disabled: boolean;
  refreshPeriod: number;
  visibility: ProjectPageVisibility;
  name: string;
  panels: Array<{ name: string; id: string }>;
  exports: Array<{ name: string; id: string }>;
}

// For exports specifically
export type Destination = { id: string } & {
  // Leaves room for other destination types in the future
  type: 'email';
  from: string;
  recipients: string;
  server: string;
  username: string;
  password_encrypt: Encrypt;
};

export class Export {
  period: 'day' | 'week' | 'month';
  disabled: boolean;
  name: string;
  id: string;
  destinations: Array<Destination>;

  constructor(defaults: Partial<Export> = {}) {
    this.period = defaults.period || 'day';
    this.name = defaults.name || 'DataStation Export';
    this.id = uuid.v4();
  }
}

export class ProjectPage {
  panels: Array<PanelInfo>;
  name: string;
  id: string;

  constructor(name?: string, panels?: Array<PanelInfo>) {
    this.name = name || '';
    this.panels = panels || [];
    this.id = uuid.v4();
  }

  static fromJSON(raw: any): ProjectPage {
    raw = raw || {};
    const pp = new ProjectPage();
    pp.panels = (raw.panels || []).map(PanelInfo.fromJSON);
    pp.name = raw.name;
    pp.id = raw.id || uuid.v4();
    return pp;
  }
}

export function doOnMatchingFields<T>(
  ps: any,
  check: (key: string) => boolean,
  cb: (f: T, path: string) => T
) {
  const stack: Array<[any, Array<string>]> = [[ps, []]];

  while (stack.length) {
    const [top, path] = stack.pop();

    if (top && typeof top === 'object') {
      for (const [elPath, el] of Object.entries(top)) {
        stack.push([el, [...path, elPath]]);
      }
    }

    const p = path.filter(Boolean).join('.');
    if (check(p)) {
      setPath(ps, p, cb(top, p));
    }
  }
}

export function doOnEncryptFields(
  ps: any,
  cb: (f: Encrypt, path: string) => Encrypt
) {
  return doOnMatchingFields(ps, (f) => f.endsWith('_encrypt'), cb);
}

export class ProjectState {
  pages: Array<ProjectPage>;
  projectName: string;
  connectors: Array<ConnectorInfo>;
  servers: Array<ServerInfo>;
  id: string;
  originalVersion: string;
  lastVersion: string;

  constructor(
    projectName?: string,
    pages?: Array<ProjectPage>,
    connectors?: Array<ConnectorInfo>,
    servers?: Array<ServerInfo>,
    originalVersion?: string,
    lastVersion?: string
  ) {
    this.pages = pages || [];
    this.projectName = projectName || '';
    this.connectors = connectors || [];
    this.servers = servers || [];
    this.originalVersion = originalVersion || VERSION;
    this.lastVersion = lastVersion || VERSION;
    this.id = uuid.v4();
  }

  static fromJSON(raw: any, external = true): ProjectState {
    raw = raw || {};
    const ps = new ProjectState();
    ps.projectName = raw.projectName || '';
    ps.pages = (raw.pages || []).map(ProjectPage.fromJSON);
    ps.connectors = (raw.connectors || []).map(ConnectorInfo.fromJSON);
    ps.servers = (raw.servers || []).map(ServerInfo.fromJSON);
    ps.id = raw.id || uuid.v4();
    ps.originalVersion = raw.originalVersion || VERSION;
    ps.lastVersion = raw.lastVersion || VERSION;
    if (external) {
      doOnEncryptFields(ps, (f, p) => {
        const new_ = new Encrypt(null);
        new_.encrypted = true;
        return new_;
      });
    }
    doOnMatchingFields<Date>(
      ps,
      (path) => path.endsWith('_date'),
      (d) => {
        const nd = new Date(d);
        if (String(nd) === 'Invalid Date') {
          return new Date();
        }

        return nd;
      }
    );
    return ps;
  }
}

export const DEFAULT_PROJECT: ProjectState = (() => {
  const page = new ProjectPage('CSV Discovery Example');

  const ppi = new ProgramPanelInfo(page.id, {
    name: 'Transform with SQL',
    type: 'sql',
    content: `SELECT name, age+5 AS age FROM DM_getPanel('Raw CSV Text');`,
  });

  const gpi = new GraphPanelInfo(page.id, { name: 'Display' });
  gpi.graph.ys = [{ field: 'age', label: 'Age' }];
  gpi.graph.x = 'name';
  gpi.graph.panelSource = ppi.id;

  page.panels = [
    new LiteralPanelInfo(page.id, {
      name: 'Raw CSV Text',
      content: 'name,age\nMorgan,12\nJames,17',
      contentTypeInfo: new ContentTypeInfo('text/csv'),
    }),
    ppi,
    gpi,
  ];

  return new ProjectState('Example project', [page]);
})();

File: runner/state.go
package runner

import (
	"fmt"
	"time"
)

type PanelResult struct {
	Exception   any      `json:"exception" db:"exception"`
	Preview     string   `json:"preview" db:"preview"`
	Stdout      string   `json:"stdout" db:"stdout"`
	Shape       Shape    `json:"shape" db:"shape"`
	ArrayCount  *float64 `json:"arrayCount" db:"arrayCount"`
	Size        *float64 `json:"size" db:"size"`
	ContentType string   `json:"contentType" db:"contentType"`
	Elapsed     *float64 `json:"elapsed" db:"elapsed"`
}

type Encrypt struct {
	Value     string `json:"value" db:"value"`
	Encrypted bool   `json:"encrypted" db:"encrypted"`
}

type ServerInfoType string

const (
	SSHAgent      = "ssh-agent"
	SSHPassword   = "password"
	SSHPrivateKey = "private-key"
)

type ServerInfo struct {
	Name           string         `json:"name" db:"name"`
	Address        string         `json:"address" db:"address"`
	Port           float64        `json:"port" db:"port"`
	Type           ServerInfoType `json:"type" db:"type"`
	Username       string         `json:"username" db:"username"`
	Password       Encrypt        `json:"password_encrypt" db:"password_encrypt"`
	PrivateKeyFile string         `json:"privateKeyFile" db:"privateKeyFile"`
	Passphrase     Encrypt        `json:"passphrase_encrypt" db:"passphrase_encrypt"`
	Id             string         `json:"id" db:"id"`
}

type ContentTypeInfo struct {
	Type             string `json:"type" db:"type"`
	CustomLineRegexp string `json:"customLineRegexp" db:"customLineRegexp"`
	ConvertNumbers   bool   `json:"convertNumbers" db:"convertNumbers"`
}

type PanelInfoType string

const (
	HttpPanel     = "http"
	ProgramPanel  = "program"
	LiteralPanel  = "literal"
	FilePanel     = "file"
	FilaggPanel   = "filagg"
	DatabasePanel = "database"
	GraphPanel    = "graph"
	TablePanel    = "table"
)

type PanelInfo struct {
	Page       int           `json:"page" db:"page"`
	PageSize   int           `json:"pageSize" db:"pageSize"`
	Content    string        `json:"content" db:"content"`
	Type       PanelInfoType `json:"type" db:"type"`
	Name       string        `json:"name" db:"name"`
	Id         string        `json:"id" db:"id"`
	ServerId   string        `json:"serverId" db:"serverId"`
	ResultMeta PanelResult   `json:"resultMeta" db:"resultMeta"`
	PageId     string        `json:"pageId" db:"pageId"`
	*ProgramPanelInfo
	*FilePanelInfo
	*LiteralPanelInfo
	*DatabasePanelInfo
	*HttpPanelInfo
	*FilaggPanelInfo
	*TablePanelInfo
	*GraphPanelInfo
}

type SupportedLanguages string

const (
	Python        SupportedLanguages = "python"
	JavaScript    SupportedLanguages = "javascript"
	Deno          SupportedLanguages = "deno"
	Ruby          SupportedLanguages = "ruby"
	R             SupportedLanguages = "r"
	Julia         SupportedLanguages = "julia"
	SQL           SupportedLanguages = "sql"
	CustomProgram SupportedLanguages = "custom"
)

type ProgramPanelInfo struct {
	Program struct {
		Type      SupportedLanguages `json:"type" db:"type"`
		CustomExe string             `json:"customExe" db:"customExe"`
	} `json:"program" db:"program"`
}

type FilePanelInfo struct {
	File struct {
		ContentTypeInfo ContentTypeInfo `json:"contentTypeInfo" db:"contentTypeInfo"`
		Name            string          `json:"name" db:"name"`
	} `json:"file" db:"file"`
}

type LiteralPanelInfoLiteral struct {
	ContentTypeInfo ContentTypeInfo `json:"contentTypeInfo" db:"contentTypeInfo"`
}

type LiteralPanelInfo struct {
	Literal LiteralPanelInfoLiteral `json:"literal" db:"literal"`
}

type HttpPanelInfo struct {
	Http HttpConnectorInfo `json:"http" db:"http"`
}

type TimeSeriesRelativeTimes string

const (
	Last5Minutes  TimeSeriesRelativeTimes = "last-5-minutes"
	Last15Minutes TimeSeriesRelativeTimes = "last-15-minutes"
	Last30Minutes TimeSeriesRelativeTimes = "last-30-minutes"
	LastHour      TimeSeriesRelativeTimes = "last-hour"
	Last3Hours    TimeSeriesRelativeTimes = "last-3-hours"
	Last6Hours    TimeSeriesRelativeTimes = "last-6-hours"
	Last12Hours   TimeSeriesRelativeTimes = "last-12-hours"
	LastDay       TimeSeriesRelativeTimes = "last-day"
	Last3Days     TimeSeriesRelativeTimes = "last-3-days"
	LastWeek      TimeSeriesRelativeTimes = "last-week"
	Last2Weeks    TimeSeriesRelativeTimes = "last-2-weeks"
	LastMonth     TimeSeriesRelativeTimes = "last-month"
	Last2Months   TimeSeriesRelativeTimes = "last-2-months"
	Last3Months   TimeSeriesRelativeTimes = "last-3-months"
	Last6Months   TimeSeriesRelativeTimes = "last-6-months"
	LastYear      TimeSeriesRelativeTimes = "last-year"
	Last2Years    TimeSeriesRelativeTimes = "last-2-years"
	AllTime       TimeSeriesRelativeTimes = "all-time"
)

type TimeSeriesFixedTimes string

const (
	ThisHour        = "this-hour"
	PreviousHour    = "previous-hour"
	Today           = "today"
	Yesterday       = "yesterday"
	WeekToDate      = "week-todate"
	PreviousWeek    = "previous-week"
	MonthToDate     = "month-to-date"
	PreviousMonth   = "previous-month"
	QuarterToDate   = "quarter-to-date"
	PreviousQuarter = "previous-quarter"
	YearToDate      = "year-to-date"
	PreviousYear    = "previous-year"
)

type TimeSeriesRangeType string

const (
	AbsoluteRange TimeSeriesRangeType = "absolute"
	RelativeRange TimeSeriesRangeType = "relative"
	FixedRange    TimeSeriesRangeType = "fixed"
	None          TimeSeriesRangeType = "none"
)

type TimeSeriesRange struct {
	Field     string                   `json:"field"`
	SortOn    *string                  `json:"sortOn"`
	SortAsc   *bool                    `json:"sortAsc"`
	Type      TimeSeriesRangeType      `json:"rangeType"`
	BeginDate *time.Time               `json:"begin_date"`
	EndDate   *time.Time               `json:"end_date"`
	Relative  *TimeSeriesRelativeTimes `json:"relative"`
	Fixed     *TimeSeriesFixedTimes    `json:"fixed"`
}

type AggregateType string

const (
	NoneAggregate    AggregateType = "none"
	CountAggregate   AggregateType = "count"
	SumAggregate     AggregateType = "sum"
	AverageAggregate AggregateType = "average"
	MinAggregate     AggregateType = "min"
	MaxAggregate     AggregateType = "max"
)

type FilaggPanelInfoFilagg struct {
	PanelSource    any             `json:"panelSource"`
	Filter         string          `json:"filter"`
	Range          TimeSeriesRange `json:"range"`
	AggregateType  AggregateType   `json:"aggregateType"`
	GroupBy        string          `json:"groupBy"`
	AggregateOn    string          `json:"aggregateOn"`
	SortOn         string          `json:"sortOn"`
	SortAsc        bool            `json:"sortAsc"`
	WindowInterval string          `json:"windowInterval"`
	Limit          int             `json:"limit"`
}

func (fpif FilaggPanelInfoFilagg) GetPanelSource() string {
	if s, ok := fpif.PanelSource.(string); ok {
		return s
	}

	return fmt.Sprintf("%v", fpif.PanelSource)
}

type FilaggPanelInfo struct {
	Filagg FilaggPanelInfoFilagg `json:"filagg"`
}

type DatabasePanelInfoDatabase struct {
	ConnectorId string            `json:"connectorId" db:"connectorId"`
	Range       TimeSeriesRange   `json:"range" db:"range"`
	Table       string            `json:"table" db:"table"`
	Step        float64           `json:"step" db:"step"`
	Extra       map[string]string `json:"extra" db:"extra"`
}

type DatabasePanelInfo struct {
	Database DatabasePanelInfoDatabase `json:"database" db:"database"`
}

type GraphPanelInfoGraph struct {
	PanelSource string        `json:"panelSource" db:"panelSource"`
	Ys          []TableColumn `json:"ys" db:"ys"`
	X           string        `json:"x" db:"x"`
}

type GraphPanelInfo struct {
	Graph GraphPanelInfoGraph `json:"graph" db:"graph"`
}

type TableColumn struct {
	Field string `json:"field"`
	Label string `json:"label"`
}

type TablePanelInfoTable struct {
	Columns     []TableColumn `json:"columns" db:"columns"`
	PanelSource string        `json:"panelSource" db:"panelSource"`
	RowNumbers  bool          `json:"rowNumbers"`
}

type TablePanelInfo struct {
	Table TablePanelInfoTable `json:"table" db:"table"`
}

type ConnectorInfoType string

const (
	DatabaseConnector ConnectorInfoType = "database"
	HTTPConnector     ConnectorInfoType = "http"
)

type ConnectorInfo struct {
	Name     string            `json:"name" db:"name"`
	Type     ConnectorInfoType `json:"type" db:"type"`
	Id       string            `json:"id" db:"id"`
	ServerId string            `json:"serverId" db:"serverId"`
	*DatabaseConnectorInfo
}

type DatabaseConnectorInfoType string

const (
	PostgresDatabase      DatabaseConnectorInfoType = "postgres"
	MySQLDatabase         DatabaseConnectorInfoType = "mysql"
	SQLiteDatabase        DatabaseConnectorInfoType = "sqlite"
	OracleDatabase        DatabaseConnectorInfoType = "oracle"
	SQLServerDatabase     DatabaseConnectorInfoType = "sqlserver"
	PrestoDatabase        DatabaseConnectorInfoType = "presto"
	ClickHouseDatabase    DatabaseConnectorInfoType = "clickhouse"
	SnowflakeDatabase     DatabaseConnectorInfoType = "snowflake"
	CassandraDatabase     DatabaseConnectorInfoType = "cassandra"
	ScyllaDatabase        DatabaseConnectorInfoType = "scylla"
	ElasticsearchDatabase DatabaseConnectorInfoType = "elasticsearch"
	SplunkDatabase        DatabaseConnectorInfoType = "splunk"
	PrometheusDatabase    DatabaseConnectorInfoType = "prometheus"
	InfluxDatabase        DatabaseConnectorInfoType = "influx"
	InfluxFluxDatabase    DatabaseConnectorInfoType = "influx-flux"
	CockroachDatabase     DatabaseConnectorInfoType = "cockroach"
	TimescaleDatabase     DatabaseConnectorInfoType = "timescale"
	CrateDatabase         DatabaseConnectorInfoType = "crate"
	YugabyteDatabase      DatabaseConnectorInfoType = "yugabyte"
	QuestDatabase         DatabaseConnectorInfoType = "quest"
	BigQueryDatabase      DatabaseConnectorInfoType = "bigquery"
	MongoDatabase         DatabaseConnectorInfoType = "mongo"
	AthenaDatabase        DatabaseConnectorInfoType = "athena"
	AirtableDatabase      DatabaseConnectorInfoType = "airtable"
	GoogleSheetsDatabase  DatabaseConnectorInfoType = "google-sheets"
	Neo4jDatabase         DatabaseConnectorInfoType = "neo4j"
	ODBCDatabase          DatabaseConnectorInfoType = "odbc"
)

type DatabaseConnectorInfoDatabase struct {
	Type     DatabaseConnectorInfoType `json:"type" db:"type"`
	Database string                    `json:"database" db:"database"`
	Username string                    `json:"username" db:"username"`
	Password Encrypt                   `json:"password_encrypt" db:"password_encrypt"`
	Address  string                    `json:"address" db:"address"`
	ApiKey   Encrypt                   `json:"apiKey_encrypt" db:"apiKey_encrypt"`
	Extra    map[string]string         `json:"extra" db:"extra"`
}

type DatabaseConnectorInfo struct {
	Database DatabaseConnectorInfoDatabase `json:"database" db:"database"`
}

type HttpConnectorInfoHeader struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type HttpConnectorInfoHttp struct {
	Method          string                    `json:"method" db:"method"`
	Url             string                    `json:"url" db:"url"`
	ContentTypeInfo ContentTypeInfo           `json:"contentTypeInfo" db:"contentTypeInfo"`
	AllowInsecure   bool                      `json:"allowInsecure" db:"allowInsecure"`
	Headers         []HttpConnectorInfoHeader `json:"headers" db:"headers"`
}

type HttpConnectorInfo struct {
	Http HttpConnectorInfoHttp `json:"http" db:"http"`
}

type ProjectPage struct {
	Panels        []PanelInfo `json:"panels" db:"panels"`
	Schedules     []any       `json:"schedules" db:"schedules"`
	Name          string      `json:"name" db:"name"`
	Id            string      `json:"id" db:"id"`
	RefreshPeriod int         `json:"refreshPeriod" db:"refreshPeriod"`
}

type ProjectState struct {
	Pages      []ProjectPage   `json:"pages" db:"pages"`
	Connectors []ConnectorInfo `json:"connectors" db:"connectors"`
	Servers    []ServerInfo    `json:"servers" db:"servers"`
	Id         string          `json:"projectName" db:"projectName"`
	// Basically never use uuid
	Uuid            string `json:"id" db:"id"`
	OriginalVersion string `json:"originalVersion" db:"originalVersion"`
	LastVersion     string `json:"lastVersion" db:"lastVersion"`
}

type CacheSettings struct {
	CachePresent bool
	Enabled      bool
}

var DefaultCacheSettings = &CacheSettings{} // Bools are set to false by default.
